modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/DistributionZoneRebalanceEngineV2.java

```diff
@@ -0,0 +1,287 @@
+                @Override
+                protected CompletableFuture<Void> onReplicasUpdate(AlterZoneEventParameters parameters, int oldReplicas) {
+                    return onUpdateReplicas(parameters);
+                }
+            });
+
+            // TODO: IGNITE-18694 - Recovery for the case when zones watch listener processed event but assignments were not updated.
+            metaStorageManager.registerPrefixWatch(zoneDataNodesKey(), dataNodesListener);
+
+            return nullCompletedFuture();
+        });
+    }
+
+    /**
+     * Stops the rebalance engine by unregistering meta storage watches.
+     */
+    public void stop() {
+        if (!stopGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        metaStorageManager.unregisterWatch(dataNodesListener);
+    }
+
+    private WatchListener createDistributionZonesDataNodesListener() {
+        return new WatchListener() {
+            @Override
+            public CompletableFuture<Void> onUpdate(WatchEvent evt) {
+                return IgniteUtils.inBusyLockAsync(busyLock, () -> {
+                    Set<Node> dataNodes = parseDataNodes(evt.entryEvent().newEntry().value());
+
+                    if (dataNodes == null) {
+                        // The zone was removed so data nodes was removed too.
+                        return nullCompletedFuture();
+                    }
+
+                    int zoneId = extractZoneIdDataNodes(evt.entryEvent().newEntry().key());
+
+                    // It is safe to get the latest version of the catalog as we are in the metastore thread.
+                    int catalogVersion = catalogService.latestCatalogVersion();
+
+                    CatalogZoneDescriptor zoneDescriptor = catalogService.zone(zoneId, catalogVersion);
+
+                    if (zoneDescriptor == null) {
+                        // Zone has been removed.
+                        return nullCompletedFuture();
+                    }
+
+                    Set<String> filteredDataNodes = filterDataNodes(
+                            dataNodes,
+                            zoneDescriptor,
+                            distributionZoneManager.nodesAttributes()
+                    );
+
+                    if (filteredDataNodes.isEmpty()) {
+                        return nullCompletedFuture();
+                    }
+
+                    return triggerPartitionsRebalanceForZone(
+                            evt.entryEvent().newEntry().revision(),
+                            zoneDescriptor,
+                            filteredDataNodes
+                    );
+                });
+            }
+
+            @Override
+            public void onError(Throwable e) {
+                LOG.warn("Unable to process data nodes event", e);
+            }
+        };
+    }
+
+    private CompletableFuture<Void> onUpdateReplicas(AlterZoneEventParameters parameters) {
+        return recalculateAssignmentsAndScheduleRebalance(
```

Comment:
rebalanceTriggersRecovery case will be covered in your next ticket, right?

Reply:
Yes, that's correct